pkg/kv/kvserver/batcheval/cmd_add_sstable.go

```diff
@@ -110,26 +110,61 @@ func EvalAddSSTable(
 	h := cArgs.Header
 	ms := cArgs.Stats
 	mvccStartKey, mvccEndKey := storage.MVCCKey{Key: args.Key}, storage.MVCCKey{Key: args.EndKey}
+	sst := args.Data
+
+	if args.Blind {
+		if args.DisallowShadowing {
+			return result.Result{}, errors.New("Blind cannot be combined with DisallowShadowing")
+		}
+		if args.IngestAsWrites {
+			return result.Result{}, errors.New("Blind cannot be combined with IngestAsWrites")
+		}
+	}
+	if args.Blind && !h.TimestampFromServerClock {
+		// FIXME: We require this in the hope that the server-assigned timestamp
+		// will be unique, such that it won't replace an existing key at the same
+		// timestamp. However, this is likely not sufficient, because even though
+		// the HLC time must be the highest seen timestamp at the time the store
+		// receives the request, other requests received later with a past timestamp
+		// (equal to ours) may still be evaluated before us. Also, this is still
+		// vulnerable to pushes, e.g. by the closed timestamp, which may not be
+		// unique.
```

Comment:
huh, I thought the tscache and latching prevented this?

Reply:
The tscache only tracks reads, not writes. And the write timestamp is allocated way before we take out latches, so that doesn't help.

A trivial example is two RPCs from nodes A and B to the leaseholder L, where the client-picked timestamps happen to be equal:

1. A sends (and commits) `Put foo@5=A bar@5=A`.
2. B sends `AddSSTable foo@5=B`.
3. A sends `Scan @5` and gets `foo@5=B bar@5=A`. What happened to its commited write `foo@5=A`?

We can try to let the server pick the timestamp, which will be highest HLC time it's seen from any request, but that's vulnerable to races:

1. B sends `AddSSTable foo@0=B`.
2. L picks the highest HLC time and updates `AddSSTable foo@7=B`.
3. Concurrently, A sends `Put foo@7=A`, which is equal to the highest HLC time L has seen.
4. L happens to schedule and evaluate `Put foo@7=A` before `AddSSTable foo@7=B` (e.g. it wins a mutex race or the concurrency manager sequences it first).
5. A sends `Get foo@5` and receives `foo@5=B`. What happened to its acknowledged write `foo@5=A`?

There are many other races too, for example with the closed timestamp which could push the `AddSSTable` timestamp up to a timestamp that someone else has already written at.

I believe the only mechanism we currently have to protect against conflicting versions is MVCC, i.e. by scanning for conflicting versions/intents/locks, and we're bypassing that mechanism with blind writes. So we need to come up with some other mechanism.

Some random off-the-cuff ideas:

* Use the current HLC time as we're evaluating the batch (after acquiring latches), which is guaranteed to be later than any other request known to the local node. This would be in the future from the batch timestamp, but the latching guards against writes at all timestamps and reads at or after the batch timestamp, so maybe it's viable.

* Add another field, e.g. `Blind`, to the MVCC timestamp which sorts before non-blind writes such that a non-blind write at the "same" timestamp would take precedence (or repurpose the `Synthetic` field for this). Of course, blind writes can then trample each other as much as they want, but maybe that's ok.

* Keep track of recent MVCC timestamps in e.g. Pebble, pick an appropriate one that's known to be unused, and take out some sort of latch on it as we're assigning the batch timestamp.

Reply:
I thought the fact that `Blind` can _overwrite_ another write was already somewhat well defined / why we only use it in narrow situations where we know it is okay. And I think all the Blind AddSSTable'ing cases we do know it is okay, because nobody else is `Put`'ing at the same time. 

But to be MVCC-ful, AddSSTable, blind or otherwise, should never be allowed to "break AOST queries and point-in-time recovery". That's the part that I thought the tscache would prevent?

Reply:
Overwriting is fine as long as it does so at a higher/different timestamp. But replacing an existing version isn't. When a client commits a value, it should be able to retrieve that value by reading at the commit timestamp. The tscache would protect against this if the committer later read that version, but if the committer only wrote it then the tscache won't protect against it and the AddSSTable can replace that version -- but the committer already knows it should exist (it was committed), so it violates ACID.

Anyway, I'm reworking this to use the eval-time HLC timestamp, which I _think_ should be fine (but needs verification). Will update the PR shortly.

Reply:
Discussed this in an [internal Slack thread](https://cockroachlabs.slack.com/archives/C021WMUML04/p1635792616011500). We cannot safely rely on the HLC to obtain an unused MVCC timestamp (due to e.g. intent resolution and future GLOBAL writes), and we don't currently have any other mechanism to do so either, so we will put the onus on the caller to maintain invariants in the case of blind writes.

I've updated the PR with what I hope will be the final form of the API, I just need to check some details and write some tests.